bazel_8: init at 8.4.1

boltzmannrain opened the pull request:
The setup is based on bazel 5,6,7 in nixpkgs but is significantly
different codeline-wise. There's a bit less patching, all patches
are via patch files and other small tweaks.

 ## Bazel 8 build
With 8.4.1 bazel dist archive is enough for air-gapped build removing
the need for FODs or `src-deps.json`, and in turn we no longer need
existing `bazel` binary to bootstrap the build.

`enableNixHacks` patch is removed, `bazel-examples` repo is used
as `passthru.tests` (cpp, java, rust) via new `bazel_8/build-support`
helpers.

`sed` patches are converted to patch files, overall amount of patching
is reduced, if there'd be need for more it can be added with links
to upstream issues or patches, ideally with a regression `passthru.tests`
item.

Amount of patching may potentially be reduced further, this needs
more feedback using this.

Overall using Bazel on NixOS remains tricky, for non-packaged repos
it may still be most practical to use FHS + extra hacks and patches,
maybe even straight up using Bazelisk and upstream pre-built binaries
in FHS.

 ## buildBazelPackage support
`pkgs/build-support/build-bazel-package/default.nix` is fairly big
with a set of tweaks that may not be necessary with recent Bazel
or may have better alternatives. More work is needed to figure
out compatibility with `bazel_8` or other things to unify with `bazel_8`
helpers.

 ## bazel_8/build-support
Here are fairly minimalistic helpers to aid building Bazel packages
using Bazel 8 features. For now only testes on `bazel-examples` repo.

`patching.nix`: just a helper to inject patches into Bazel external
dependencies

`bazelDerivation`: helps to compose Bazel commandline and consume
`repository_cache`, `vendor_dir`, `registry` all related to external
dependencies management

`bazelPackage`: two-stage derivation similar to `buildBazelPackage`,
having options for `repository_cache` (pure fetches) or `vendor_dir`
(unpacked fetches, further patchable)
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

arximboldi commented:
Reporting that I've tried this and it seems to be working pretty well!

Ping @NixOS/bazel as this seems to have gone unnoticed

ryand56 commented:
Any progress on this? Or just waiting to be merged?

timothyklim commented:
8.4.0 will release next week but nixpkgs doesn't have even 8.0 release =\

hzeller commented:
I get the impression the original @NixOS/bazel  maintainers moved on to other projects ?

Anyway, for anyone with a commit bit: I think getting this in now would be great; if there are things that need fixing later, it is then possible to iterate from there.

phaer commented:
@hzeller Right. I am not a bazel user myself, but the lack of a response by maintainer, the PRs age and the number of positive reviews and upvotes make me believe we are better of merging this.

It would be nice if one of you could look into the failing test on x86_64-darwin.

boltzmannrain commented:
@ofborg build bazel_8

emilazy reviewed (commented):
Thanks for your work on this!

@boltzmannrain @hzeller Since the Bazel team appears to be inactive, how do you feel about adding yourselves to it (and considering dropping inactive maintainers from it)? The Bazel support in Nixpkgs could clearly use a lot of work and it would be good to have an active maintainer team for that. Being part of the team will also let you use the merge bot to merge PRs opened by @r-ryantm and by committers (though I imagine getting @r-ryantm to create automatic update PRs for this would require some `passthru.updateScript` engineering and some good fortune).

I realize that a committer to merge the changes is the bigger blocker here. I do not understand Bazel enough to feel qualified to give a truly thorough review of this, but the code does not scare me any more than the existing Bazel derivation, so I’m okay merging this once builds are verified. An update to the latest version would be great, but can come in a follow‐up PR. FWIW, it would be very nice to see this package become version‐generic rather than copy‐pasting, but it seems like the Bazel 8 build is different enough that a fresh start makes sense.

> Hmm, excluding darwin from platforms didn't help with CI wait time, check on darwin is still pending. Maybe because it was previously requested on this PR?

ofborg will still try to queue things up on Darwin even if they’re marked as unsupported, because it only finds out that the package is unsupported once the queue entry runs. Unfortunately ofborg hasn’t gotten as much attention as it needs lately and the Darwin runners in particular are very overloaded. I can help verify a Darwin build if you remove the exclusion commit. You can ping `@NixOS/darwin-maintainers` for that kind of thing in future, too.

boltzmannrain commented on pkgs/by-name/ba/bazel_8/package.nix:
```diff
@@ -0,0 +1,352 @@
+  file,
+  installShellFiles,
+  python3,
+  # Apple dependencies
+  cctools,
```
I have access to darwin machines, but sometimes the problem is failures only happening in ofborg (maybe I should spin up ofborg too, in case machine/container setup is reproducible?)

boltzmannrain replied:
cctools.libtool works, applied the suggestion now

boltzmannrain committed:
bazel_8: init at 8.4.1

The setup is based on bazel 5,6,7 in nixpkgs but is significantly
different codeline-wise. There's a bit less patching, all patches
are via patch files and other small tweaks.

 ## Bazel 8 build
With 8.4.1 bazel dist archive is enough for air-gapped build removing
the need for FODs or `src-deps.json`, and in turn we no longer need
existing `bazel` binary to bootstrap the build.

`enableNixHacks` patch is removed, `bazel-examples` repo is used
as `passthru.tests` (cpp, java, rust) via new `bazel_8/build-support`
helpers.

`sed` patches are converted to patch files, overall amount of patching
is reduced, if there'd be need for more it can be added with links
to upstream issues or patches, ideally with a regression `passthru.tests`
item.

Amount of patching may potentially be reduced further, this needs
more feedback using this.

Overall using Bazel on NixOS remains tricky, for non-packaged repos
it may still be most practical to use FHS + extra hacks and patches,
maybe even straight up using Bazelisk and upstream pre-built binaries
in FHS.

 ## buildBazelPackage support
`pkgs/build-support/build-bazel-package/default.nix` is fairly big
with a set of tweaks that may not be necessary with recent Bazel
or may have better alternatives. More work is needed to figure
out compatibility with `bazel_8` or other things to unify with `bazel_8`
helpers.

 ## bazel_8/build-support
Here are fairly minimalistic helpers to aid building Bazel packages
using Bazel 8 features. For now only testes on `bazel-examples` repo.

`patching.nix`: just a helper to inject patches into Bazel external
dependencies

`bazelDerivation`: helps to compose Bazel commandline and consume
`repository_cache`, `vendor_dir`, `registry` all related to external
dependencies management

`bazelPackage`: two-stage derivation similar to `buildBazelPackage`,
having options for `repository_cache` (pure fetches) or `vendor_dir`
(unpacked fetches, further patchable)

boltzmannrain commented on pkgs/top-level/all-packages.nix:
```diff
@@ -6581,6 +6581,11 @@ with pkgs;
 
+  bazel_8 = callPackage ../by-name/ba/bazel_8/package.nix {
+    # https://github.com/bazelbuild/bazel/issues/25124
+    stdenv = if stdenv.cc.isClang then llvmPackages_17.stdenv else stdenv;
```
Turns out it's even better, 8.4.0 works without any workarounds, removed the overrides now

boltzmannrain commented:
Updated to use 8.4.0, this allowed to remove couple more patches. Applied other suggestions from comments here and from Gemini code assist from https://github.com/boltzmannrain/nixpkgs/pull/1

Keeping platforms = linux + darwin, on my machines all are building fine, and removing won't help skip the darwin queue anyway.

donn commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 400941`
Commit: `fae41a5ee90b944d3bb461038ccb0fb168373ea6`

---
### `x86_64-darwin`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>bazel_8</li>
  </ul>
</details>

---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>bazel_8</li>
  </ul>
</details>

emilazy reviewed (approved):
Thanks for sticking with this work! Hope this can help improve the Bazel + Nix situation :)
